feat(client): support session autosaves

lorenzo-cavazzi opened the pull request:
This PR implements most of the changes from #1345 , except the visual changes to the commits info in the dropdown menu.
There are a few additional changes to simplify selecting an already running session for a target branch.

In short, these are the changes:
* Get and delete autosaves.
* Select new session commit based on autosaves and running sessions.
* Update CheckNotebookStatus component to support per-branch sessions.

**How to test**
Start a session on any project (possibly from last - 1 or -2 commit), do some work, and stop the session.
You should now have an autosave. If you try to start a session again, you should see some info when trying to select a branch/commit (P.S. this is silent unless you click on "select branch/commit/..." -- see screenshots below).
Notice that the older commits should be selected when
- No session is running, but the autosave is detected.
- The session is already running (take priority over the last commit)
- File preview on notebooks should also detect running sessions form older commits

**Screenshots**
* Autosave content available
![Screenshot_20220117_154255](https://user-images.githubusercontent.com/43481553/149918903-90e4f789-3c97-4377-a7bc-e670fac71681.png)

* User trying to start from another commit
![Screenshot_20220117_154313](https://user-images.githubusercontent.com/43481553/149918960-19487279-8cd5-44b2-b510-8d8cf9e481fc.png)

* Error on delete --> we don't get any info about the error, but it may (rarely) happen if the user has already "consumed" the autosave in another tab and left this open 
![Screenshot_20220117_134806](https://user-images.githubusercontent.com/43481553/149919016-9ab1dba1-8580-46a8-a999-d19cd8c9a99f.png)



/deploy renku=1.0-next renku-graph=cli-v1 renku-core=0000-disbale-check-migration-optimization
fix #1345, #789

ciyer reviewed (changes requested):
Nice feature which I think simplifies and improves the handling of autosave information! 🎉 

After trying it out, I think we should make some additional simplifications to the UI to make it even easier to use.

1. If there is an autosave for a particular branch, and the user wants to start a session on that branch, then force the user to either start on the commit with the autosave information or delete the autosave.

Right now, if the user starts on a different commit, the autosave information is removed for them; this makes that step more explicit and reduces opportunities for accidental data loss, I think.

In that case, we can change the message to be:
```
There is unsaved work from your last session which will be restored. If you do not wish to keep it, 
you can [delete the unsaved work](invoke delete command).

For more options, start a session and look at the session cheatsheet, which is available under this icon [show the session cheatsheet icon].
```

Add the following to the session cheatsheet:

## Autosave

If you see what work was saved you can diff the current contents against the previous state:

    git diff HEAD^

or against the latest state of the project:

    git diff origin [default branch name]

If you want to discard the unsaved work in a running session and return to the latest version:

    git reset --hard origin/[default branch name]

If you would like to keep this work around for later, but continue working with the latest version for now:

    git push --set-upstream origin your-branch-name-here
    git reset --hard origin/[default branch name]

## Autosave information not being correctly updated

I had an autosave branch, which I removed by starting from another commit. However, I still get this message, even though the autosave branch is gone.

![image](https://user-images.githubusercontent.com/1196411/149951935-06106868-bd45-4da3-975a-914e515b9c20.png)


![image](https://user-images.githubusercontent.com/1196411/149952225-1fe8465d-a8c8-4a4d-b35f-dde2cfe2fa72.png)

lorenzo-cavazzi commented:
I've added the autosave info on the Cheatsheet (I slightly changed the words to use the 2 column design, but I can switch back to the linear one and use your suggestion verbatim), changed the alert info, and fixed the bug resulting in the autosave list not being cleaned when the last available autosave is deleted.

![Screenshot_20220119_144658](https://user-images.githubusercontent.com/43481553/150143535-f73fc459-3e37-4dee-a575-e9a836e6871d.png)

ciyer reviewed (changes requested):
This is very close now. I just added a commit to clean up the cheat sheet a bit. I also found an issue with the `faBook` icon, which should be easy to fix.

I want to understand the situations where the Autosave service can return an error, but that is more for my knowledge.

lorenzo-cavazzi commented on client/src/notebooks/Notebooks.present.js:
```diff
@@ -1068,74 +1068,157 @@ function StartNotebookServer(props) {
+        </p>
+        <p className="mb-0">
+          For more options, start a session and look at the session cheatsheet,
+          which is available under this icon <FontAwesomeIcon icon={faBook} />.
```
Ah true, thanks for spotting that!

lorenzo-cavazzi commented:
> I want to understand the situations where the Autosave service can return an error

This happens when the page to start a new session is already open and the autosave banner is visible, but the autosave gets removed from somewhere else (E.G. the branch is manually removed from GitLab, or the user starts a session on that branch from a different commit on another tab).
If the user tries to remove the autosave, then the `DELETE /autosave` API will return an error.

olevski commented:
@lorenzo-cavazzi this looks good to me. I tested a few cases.

One thing that I saw is that if you make a new branch but do not push it, then the autosave branch gets properly created but it is impossible to get the data out of it. This is because there is still no such branch in the remote repo. And the newly created branch does not show up in the list of branches to select when starting a session.

This is not a result of this PR though. It is just a problem with how the autosave system currently works. I will think of how we can tackle this and open a new issue for it somewhere. Let me know if you have any ideas.

lorenzo-cavazzi commented:
Thanks for the feedback @olevski !
I'm wondering if using the starting branch for generating the autosave instead of the current branch would be a solution. But maybe it's not the way to go. It may be very confusing to start a session on a branch and end up in a different one.

Also, the UI will switch at some point to a per-branch approach, where the user can select/change the branch when browsing a project, and the UI will show the related content. There it will be even more difficult to identify a non-pushed branch.

I would call this a corner case though. Autosaves should prevent data loss, but branching off and not pushing may be a bit of a misuse of autosaves. And probably not a very common case.

olevski commented:
So yeah this issue is good to go as far as I am concerned. In terms of the additional stuff I found here is an issue that would address it on the notebook side. https://github.com/SwissDataScienceCenter/renku-notebooks/issues/884.

I think that the pre-stop hook that creates the autosave branch can check for this specific weird situation. And if we are in this situation when the user just created a new branch without publishing it then the pre-stop script can simply publish the branch and allow the user to then recover from it later on.
